source/control/daq_control.hh

```diff
@@ -2,167 +2,55 @@
-            };
-            class status_error : public error
-            {
-                public:
-                    status_error() {}
-                    virtual ~status_error() {}
-            };
 
         public:
-            daq_control( const scarab::param_node& a_master_config, std::shared_ptr< stream_manager > a_mgr );
+            daq_control( const scarab::param_node& a_primary_config, std::shared_ptr< sandfly::stream_manager > a_mgr );
             virtual ~daq_control();
 
-            /// Pre-execution initialization (call after setting the control_access pointer)
-            void initialize();
-
-            /// Run the DAQ control thread
-            void execute( std::condition_variable& a_ready_condition_variable, std::mutex& a_ready_mutex );
-
-            /// Start the DAQ into the activated state
-            /// Can throw daq_control::status_error; daq_control will still be usable
-            /// Can throw psyllid::error; daq_control will NOT be usable
-            void activate();
-            /// Restarts the DAQ into the activated state
-            /// Can throw daq_control::status_error; daq_control will still be usable
-            /// Can throw psyllid::error; daq_control will NOT be usable
-            void reactivate();
-            /// Returns the DAQ to the deactivated state
-            /// Can throw daq_control::status_error; daq_control will still be usable
-            /// Can throw psyllid::error; daq_control will NOT be usable
-            void deactivate();
-
-            bool is_ready_at_startup() const;
-
-            /// Start a run
-            /// Can throw daq_control::run_error or status_error; daq_control will still be usable
-            /// Can throw psyllid::error; daq_control will NOT be usable
-            /// Stop run with stop_run()
-            void start_run();
-
-            /// Stop a run
-            /// Can throw daq_control::run_error or status_error; daq_control will still be usable
-            /// Can throw psyllid::error; daq_control will NOT be usable
-            void stop_run();
-
-        public:
-            void apply_config( const std::string& a_node_name, const scarab::param_node& a_config );
-            void dump_config( const std::string& a_node_name, scarab::param_node& a_config );
+        protected:
+            /// Handle called to perform initialization
+            virtual void on_initialize();
+            /// Handle called before run takes place (i.e. control is unpaused)
+            virtual void on_pre_run();
+            /// Handle called after run finishes (i.e. control is paused)
+            virtual void on_post_run();
 
-            /// Instruct a node to run a command
-            /// Throws psyllid::error if the command fails; returns false if the command is not recognized
-            bool run_command( const std::string& a_node_name, const std::string& a_cmd, const scarab::param_node& a_args );
 
         public:
```

Comment:
The class is a bit messy, the public keyword appears multiple times for no reason and some functions are declared multiple times.

Reply:
I agree about the extra `public` keyword.  I'm not sure what functions are declared multiple times.

Reply:
You are right, sorry about that. I think I was comparing the two files with vimdiff, which probably made it more confusing.